test/deadline_submitter_for_houdini/unit/test_job_template.py

```diff
@@ -0,0 +1,187 @@
+# Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+
+from unittest.mock import Mock, patch
+import pytest
+from deadline.houdini_submitter.python.deadline_cloud_for_houdini.submitter import (
+    _get_job_template,
+    RenderStrategy,
+)
+
+
+@pytest.fixture
+def mock_parm():
+    def mock_implementation(name: str):
+        if name == "name":
+            mock_name = Mock()
+            mock_name.evalAsString.return_value = "My Job"
+            return mock_name
+        if name == "include_adaptor_wheels":
+            mock_name = Mock()
+            mock_name.eval.return_value = False
+            return mock_name
+        if name == "description":
+            mock_name = Mock()
+            mock_name.evalAsString.return_value = "My job description"
+            return mock_name
+        return Mock()
+
+    return mock_implementation
+
+
+@patch(
+    "deadline.houdini_submitter.python.deadline_cloud_for_houdini.submitter._get_houdini_version",
+    Mock(return_value="19.5.435"),
+)
+@patch(
+    "deadline.houdini_submitter.python.deadline_cloud_for_houdini.submitter._get_hip_file",
+    Mock(return_value="/path/to/hip.hip"),
+)
+@patch("deadline.houdini_submitter.python.deadline_cloud_for_houdini.submitter._get_steps")
+def test_job_template(mock_get_steps, mock_parm):
+    mock_node = Mock()
+    mock_node.userData.return_value = None
+    mock_node.name = "Test Job"
+    mock_node.parm = mock_parm
+
+    mock_get_steps.return_value = [
+        {
+            "id": "1",
+            "name": "/mantra-1",
+            "dependency_ids": [],
+            "rop": "/mantra",
+            "wedgenum": "",
+            "wedge_node": "",
+            "start": 1,
+            "end": 5,
+            "step": 1,
+            "render_strategy": RenderStrategy.PARALLEL,
+        }
+    ]
+    template = _get_job_template(mock_node)
```

Comment:
I think the functions organize and encapsulate different bits of logic reasonably well, and the tests run them through the various cases we care about. Is the private vs public distinction the main thing you're looking at? What do you think about just making the functions public?

Reply:
I looked more at making these public functions, but I think they made more sense as private functions since they're only used within this file. Another pattern we could do is move them into a helpers/utils file and make them public, but that would just spread out the code and IMO would make it less readable. I'm going to leave them as private for now but am open to reorganizing it if you have specific ideas.

Reply:
Yeah, it's generally an anti-pattern to test (and mock) private functions as it usually means that we are testing implementation details rather than behaviour. Sometimes it is reasonable to test private functions given constraints/time/code structure, but it could also indicate opportunities to refactor the code for better readability, SOLID principles, etc. In this case, I don't think it's a blocker but could be an opportunity for future refactors 🙂

Reply:
Something to keep in mind here is that a public interface for applications is different than the interface for applications and we're not reflecting that correctly in some of these repos.

Effectively, the only public thing should be the part that actually hooks up the submitter with the DCC. Everything else internally is effectively private, but we've done this at the individual function level rather than the module level.